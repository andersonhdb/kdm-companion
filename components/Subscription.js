import React from 'react'
import {
  Button,
  Caption,
  Divider,
  Icon,
  Image,
  Row,
  Spinner,
  Subtitle,
  Text,
  Title,
  View,
} from '@shoutem/ui'
import { Alert, Platform, NativeModules } from 'react-native'
import * as RNIap from 'react-native-iap'
import { observer, inject } from 'mobx-react/native'

import colors from '../src/colors'

function hasIAP() {
  return !!NativeModules.RNIapModule
}

@inject(({ store }) => ({
  subscription: store.subscription,
}))
@observer
export class SubscriptionUpdater extends React.Component {
  state = {
    connected: false,
  }

  checkPurchases = async () => {
    if (!this.state.connected) {
      return
    }

    try {
      const purchases = await RNIap.getAvailablePurchases()
      console.log('purchases', purchases)
      if (purchases.length === 0) {
        // no active subscriptions
        this.props.subscription.purchased({})
      }
      purchases.forEach(purchase => {
        this.props.subscription.purchased(purchase)
      })
    } catch (err) {
      if (err.code === 'E_NOT_PREPARED') {
        // lost connection, try to re-connect the next time.
        this.setState({ connected: false })
      }
      console.warn(err.code, err.message)
    }
  }

  connectIAP = async () => {
    if (this.state.connected) {
      return
    }

    try {
      await RNIap.initConnection()
      this.setState({ connected: true })
    } catch (err) {
      if (err.code === 'E_NOT_ENDED') {
        // already connected, ignore.
        this.setState({ connected: true })
      } else {
        // 'E_SERVICE_ERROR' => store is not available.
        console.warn(err.code, err.message)
      }
    }
  }

  async componentDidMount() {
    if (hasIAP()) {
      this.connectIAP()
      this.checkPurchases() // check right away.
    }
  }

  async componentWillUnmount() {
    if (hasIAP()) {
      try {
        await RNIap.endConnection()
      } catch (err) {
        console.warn(err.code, err.message)
      }
    }
  }

  async componentDidUpdate(prevProps, prevState, snapshot) {
    if (hasIAP() && this.props.subscription.shouldCheck()) {
      await this.connectIAP()
      this.checkPurchases()
    }
  }

  render() {
    return this.props.children
  }
}

@inject(({ store }) => ({
  subscription: store.subscription,
}))
@observer
export default class Subscription extends React.Component {
  state = {
    waiting: false,
    //
    // example: [{
    //   introductoryPricePeriodAndroid: '',
    //   introductoryPriceCyclesAndroid: '',
    //   productId: 'premium_subscription_1',
    //   price: '0.99',
    //   currency: 'USD',
    //   type: 'subs',
    //   localizedPrice: '$0.99',
    //   freeTrialPeriodAndroid: 'P5D',
    //   title: 'Premium Subscription (Kingdom Death: Monster Companion)',
    //   description: 'Access premium features of the app',
    //   introductoryPrice: '',
    //   subscriptionPeriodAndroid: 'P1M',
    // }],
    products: [],
    error: false,
    showSubscriptionButton: false,
    connected: false,
  }

  connectIAP = async () => {
    if (this.state.connected) {
      return true
    }

    let status = true

    try {
      await RNIap.initConnection()
      this.setState({ connected: true })
    } catch (err) {
      if (err.code === 'E_NOT_ENDED') {
        // already connected, ignore.
        this.setState({ connected: true })
      } else {
        // 'E_SERVICE_ERROR' => store is not available.
        console.warn(err.code, err.message)
        status = false
      }
    }
    return status
  }

  getSubscriptions = async () => {
    if (!this.state.connected) {
      return true
    }

    const itemSkus = Platform.select({
      ios: ['premium_subscription_1'],
      android: ['premium_subscription_1'],
    })
    let products = []
    try {
      products = await RNIap.getSubscriptions(itemSkus)
    } catch (err) {
      if (err.code === 'E_NOT_PREPARED') {
        // lost connection, try to re-connect the next time.
        this.setState({ connected: false, error: true })
      } else {
        console.warn(err.code, err.message)
      }
    }
    return products
  }

  buySubscribeItem = async sku => {
    if (!this.state.connected) {
      return true
    }

    try {
      this.setState({ error: false, waiting: true })
      const purchase = await RNIap.buySubscription(sku)
      console.log('purchase', purchase)
      if (purchase.transactionId) {
        this.props.subscription.purchased(purchase)
      }
    } catch (err) {
      if (err.code === 'E_USER_CANCELLED') {
        // 'E_USER_CANCELLED' => user closed the window without completing purchase
      } else if (err.code === 'E_NOT_PREPARED') {
        // lost connection, try to re-connect the next time.
        this.setState({ connected: false, error: true })
      } else if (err.code === 'E_ALREADY_OWNED') {
        // user already has the subscription, but we don't have it in our side :/, let's synch the purchases.
        await this.checkPurchases()
      } else {
        console.warn(err.code, err.message)
        // 'E_UNKNOWN' => CC fails ?
        this.setState({ error: true })
      }
    }
    this.setState({ waiting: false })
  }

  checkPurchases = async () => {
    if (!this.state.connected) {
      return
    }

    try {
      const purchases = await RNIap.getAvailablePurchases()
      console.log('purchases', purchases)
      if (purchases.length === 0) {
        // no active subscriptions
        this.props.subscription.purchased({})
      }
      purchases.forEach(purchase => {
        this.props.subscription.purchased(purchase)
      })
    } catch (err) {
      if (err.code === 'E_NOT_PREPARED') {
        // lost connection, try to re-connect the next time.
        this.setState({ connected: false })
      }
      console.warn(err.code, err.message)
    }
  }

  async componentDidMount() {
    if (hasIAP() && !this.props.subscription.hasActiveSubscription()) {
      let result = await this.connectIAP()
      if (!result) {
        this.setState({ waiting: false, error: true })
      } else {
        let products = await this.getSubscriptions()
        console.log('products', products)
        let error = products.length === 0
        this.setState({ products, waiting: false, error })
      }
    }
  }

  async componentDidUpdate(prevProps, prevState, snapshot) {
    if (
      hasIAP() &&
      !this.props.subscription.hasActiveSubscription() &&
      !this.state.connected
    ) {
      this.connectIAP()
    }
  }

  async componentWillUnmount() {
    if (hasIAP()) {
      try {
        await RNIap.endConnection()
      } catch (err) {
        if (err.code === 'endConnection') {
          // already ended, ignore.
        } else {
          console.warn(err.code, err.message)
        }
      }
    }
  }

  subscribed() {
    return (
      <Text>
        Thanks for your support! you are already subscribed to Premium.
      </Text>
    )
  }

  unsubscribed() {
    return (
      <View>
        <Text>This feature is only enabled for Premium users.</Text>
        <Text>
          By becoming a Premium subscriber you can unlock advanced features of
          the app and also you will be supporting further development of the
          app.
        </Text>
        <Divider />
        {this.state.error && (
          <View>
            <Text style={styles.error}>
              An error ocurred contacting the store.
            </Text>
            <Text style={styles.error}>Please try again later.</Text>
          </View>
        )}

        {this.state.products.length > 0 && (
          <Button
            onPress={() =>
              this.buySubscribeItem(this.state.products[0].productId)
            }
          >
            {this.state.waiting && <Spinner style={styles.buttonSpinner} />}
            <Text>Subscribe for {this.state.products[0].localizedPrice}</Text>
          </Button>
        )}
      </View>
    )
  }

  render() {
    let text = this.props.subscription.hasActiveSubscription()
      ? this.subscribed()
      : this.unsubscribed()
    return (
      <View>
        <Title>Premium Subscription</Title>
        {text}
        {this.state.waiting &&
          this.state.products.length === 0 && (
            <View styleName="horizontal">
              <Text>Connecting to the store</Text>
              <Spinner />
            </View>
          )}
      </View>
    )
  }
}

const styles = {
  modal: {
    backgroundColor: colors.grey900,
    padding: 8,
  },
  error: {
    color: colors.red900,
  },
  buttonSpinner: {
    color: colors.grey900,
  },
}
