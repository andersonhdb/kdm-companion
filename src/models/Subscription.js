import { types } from 'mobx-state-tree'
import Constants from 'expo-constants'

const toDays = 1000 * 60 * 60 * 24

export const checkEvery = 1 * toDays // 1000 * 60 * 7
export const activeGracePeriod = 31 * toDays // since last check

const Subscription = types
  .model('Subscription', {
    transactionId: types.maybe(types.string), // id of the subscription received.
    productId: '', // product sku purchased.
    response: '', // full json response (in case we need somehting else)
    purchasedAt: 0, // timestamp the subscription was purchased
    autoRenewing: false, // if the subscription is autoRenewing, false => is not going to review, but it might still be active.
    checkedAt: 0, // timestamp when was it last checked if the subscription is still active, this should trigger a re-check after some time to make sure we de-active folks that cancelled their subscription.
    appLastActiveAt: 0, // timestamp when the app was last active (eg: context switched or opened), this is used so that we can re-check if the app is never closed.
  })
  .actions(self => ({
    // when subscription is still active and going to be renewed, the subscription is returned and autoRenewing = true, note that date of purchase is always the first one.
    // when subscription is still active and not going to be renewed, the subscription is returned and autoRenewing = false. Ideally a warning should be displayed in the app and a button to renew the subscription.
    // when subscription already expired, an empty response is returned.
    purchased(response) {
      if (!response || !response.transactionId) {
        self.response = ''
        self.transactionId = undefined
        self.productId = ''
        self.autoRenewing = false
        self.purchasedAt = 0
        self.checkedAt = Date.now()
        return
      }

      let purchasedAt = Number.parseInt(response.transactionDate)
      // only keep the latest purchase
      if (purchasedAt < self.purchasedAt) {
        return
      }

      self.response = JSON.stringify(response)
      self.transactionId = response.transactionId
      self.productId = response.productId
      self.autoRenewing = response.autoRenewingAndroid
      self.purchasedAt = purchasedAt
      self.checkedAt = Date.now()
    },
    updateAppLastActiveAt() {
      self.appLastActiveAt = Date.now()
    },
    // these are actions instead of view to avoid memoization (because Date is used.)
    shouldCheck() {
      if (
        self.hasActiveSubscription() &&
        Date.now() - self.checkedAt >= checkEvery
      ) {
        return true
      } else {
        return false
      }
    },
    hasActiveSubscription() {
      if (
        Constants &&
        Constants.manifest &&
        (Constants.manifest.releaseChannel === undefined ||
          Constants.manifest.releaseChannel === 'default')
      ) {
        return true
      }
      if (self.purchasedAt && Date.now() - self.checkedAt < activeGracePeriod) {
        return true
      }
      return false
    },
  }))
export { Subscription }

// log of check subscription responses
// initial check after a successful subscription
// [10:05:17] purchases Array [
// [10:05:17]   Object {
// [10:05:17]     "autoRenewingAndroid": true,
// [10:05:17]     "dataAndroid": "{\"orderId\":\"GPA.3394-3697-8517-09181\",\"packageName\":\"com.github.ingoclaro.kdmcompanion\",\"productId\":\"premium_subscription_1\",\"purchaseTime\":1536944678259,\"purchaseState\":0,\"purchaseToken\":\"elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ\",\"autoRenewing\":true}",
// [10:05:17]     "productId": "premium_subscription_1",
// [10:05:17]     "purchaseToken": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:05:17]     "signatureAndroid": "jJWtCsOSWp355YWNVUClS+wKS+OnezmCsi8L4lWYfZrmpqlm6JFYe7A9iLbTAQLpHY+HaOKOli/EnRgLF82grElDs/DE4ujbARqdXCz2d2Dp6a7vVsQhJLYF3gSLTEgKFDXWIBtnRY1FfocF5CEgiuZ+5pF6keTmQJk9/M5RY5xew408jvSpDfP8haC2jxanPAL6mZ9lJZTOr1lsNhXOtJRJ7BvCRZlXBQYThsaZKt5qBRp9btApq4PlkI8mUCuaKwD5j3kXbZw9ZLHVN1i0guj76+yi54EExjAZ/fitRrjFvLxWo0aZJFqgWtVQ/bXGyw5ZuANr5S2GhxG1JntgkA==",
// [10:05:17]     "transactionDate": "1536944678259",
// [10:05:17]     "transactionId": "GPA.3394-3697-8517-09181",
// [10:05:17]     "transactionReceipt": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:05:17]   },
// [10:05:17] ]
// in playstore UI: subscriber since: Sep 14, 2018, 10:04 AM PDT. Renews on: Sep 14, 2018, 10:09 AM PDT.

// after auto renewal (note that json response didn't change):
// [10:10:14] purchases Array [
// [10:10:14]   Object {
// [10:10:14]     "autoRenewingAndroid": true,
// [10:10:14]     "dataAndroid": "{\"orderId\":\"GPA.3394-3697-8517-09181\",\"packageName\":\"com.github.ingoclaro.kdmcompanion\",\"productId\":\"premium_subscription_1\",\"purchaseTime\":1536944678259,\"purchaseState\":0,\"purchaseToken\":\"elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ\",\"autoRenewing\":true}",
// [10:10:14]     "productId": "premium_subscription_1",
// [10:10:14]     "purchaseToken": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:10:14]     "signatureAndroid": "jJWtCsOSWp355YWNVUClS+wKS+OnezmCsi8L4lWYfZrmpqlm6JFYe7A9iLbTAQLpHY+HaOKOli/EnRgLF82grElDs/DE4ujbARqdXCz2d2Dp6a7vVsQhJLYF3gSLTEgKFDXWIBtnRY1FfocF5CEgiuZ+5pF6keTmQJk9/M5RY5xew408jvSpDfP8haC2jxanPAL6mZ9lJZTOr1lsNhXOtJRJ7BvCRZlXBQYThsaZKt5qBRp9btApq4PlkI8mUCuaKwD5j3kXbZw9ZLHVN1i0guj76+yi54EExjAZ/fitRrjFvLxWo0aZJFqgWtVQ/bXGyw5ZuANr5S2GhxG1JntgkA==",
// [10:10:14]     "transactionDate": "1536944678259",
// [10:10:14]     "transactionId": "GPA.3394-3697-8517-09181",
// [10:10:14]     "transactionReceipt": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:10:14]   },
// [10:10:14] ]
// in playstore UI: subscriber since: Sep 14, 2018, 10:04 AM PDT. Renews on: Sep 14, 2018, 10:14 AM PDT.

// after auto renewal (note that json response didn't change):
// [10:15:06] purchases Array [
// [10:15:06]   Object {
// [10:15:06]     "autoRenewingAndroid": true,
// [10:15:06]     "dataAndroid": "{\"orderId\":\"GPA.3394-3697-8517-09181\",\"packageName\":\"com.github.ingoclaro.kdmcompanion\",\"productId\":\"premium_subscription_1\",\"purchaseTime\":1536944678259,\"purchaseState\":0,\"purchaseToken\":\"elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ\",\"autoRenewing\":true}",
// [10:15:06]     "productId": "premium_subscription_1",
// [10:15:06]     "purchaseToken": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:15:06]     "signatureAndroid": "jJWtCsOSWp355YWNVUClS+wKS+OnezmCsi8L4lWYfZrmpqlm6JFYe7A9iLbTAQLpHY+HaOKOli/EnRgLF82grElDs/DE4ujbARqdXCz2d2Dp6a7vVsQhJLYF3gSLTEgKFDXWIBtnRY1FfocF5CEgiuZ+5pF6keTmQJk9/M5RY5xew408jvSpDfP8haC2jxanPAL6mZ9lJZTOr1lsNhXOtJRJ7BvCRZlXBQYThsaZKt5qBRp9btApq4PlkI8mUCuaKwD5j3kXbZw9ZLHVN1i0guj76+yi54EExjAZ/fitRrjFvLxWo0aZJFqgWtVQ/bXGyw5ZuANr5S2GhxG1JntgkA==",
// [10:15:06]     "transactionDate": "1536944678259",
// [10:15:06]     "transactionId": "GPA.3394-3697-8517-09181",
// [10:15:06]     "transactionReceipt": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:15:06]   },
// [10:15:06] ]
// in playstore UI: subscriber since: Sep 14, 2018, 10:04 AM PDT. Renews on: Sep 14, 2018, 10:19 AM PDT.

// changing to failed CC, after renewal (note that json response didn't change):
// [10:20:33] purchases Array [
// [10:20:33]   Object {
// [10:20:33]     "autoRenewingAndroid": true,
// [10:20:33]     "dataAndroid": "{\"orderId\":\"GPA.3394-3697-8517-09181\",\"packageName\":\"com.github.ingoclaro.kdmcompanion\",\"productId\":\"premium_subscription_1\",\"purchaseTime\":1536944678259,\"purchaseState\":0,\"purchaseToken\":\"elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ\",\"autoRenewing\":true}",
// [10:20:33]     "productId": "premium_subscription_1",
// [10:20:33]     "purchaseToken": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:20:33]     "signatureAndroid": "jJWtCsOSWp355YWNVUClS+wKS+OnezmCsi8L4lWYfZrmpqlm6JFYe7A9iLbTAQLpHY+HaOKOli/EnRgLF82grElDs/DE4ujbARqdXCz2d2Dp6a7vVsQhJLYF3gSLTEgKFDXWIBtnRY1FfocF5CEgiuZ+5pF6keTmQJk9/M5RY5xew408jvSpDfP8haC2jxanPAL6mZ9lJZTOr1lsNhXOtJRJ7BvCRZlXBQYThsaZKt5qBRp9btApq4PlkI8mUCuaKwD5j3kXbZw9ZLHVN1i0guj76+yi54EExjAZ/fitRrjFvLxWo0aZJFqgWtVQ/bXGyw5ZuANr5S2GhxG1JntgkA==",
// [10:20:33]     "transactionDate": "1536944678259",
// [10:20:33]     "transactionId": "GPA.3394-3697-8517-09181",
// [10:20:33]     "transactionReceipt": "elnaofdgfkndkkdbpdmplppl.AO-J1OzCazlSEK0i7DtQVeEXxDYHgsyqjDKJMC9hZjMwyEKCb2uYxjG9Muamb8bdv-5lSs1FyRRr7p8O15JLV70-Ujit8-ywejmFJcWng5aVvqk5nU08yA0nGgBoZuABJxYs7NVTNRWF6nZAs0HP_zrp1DDbD7rijQ",
// [10:20:33]   },
// [10:20:33] ]
// in playstore UI: subscriber since: Sep 14, 2018, 10:04 AM PDT. Renews on: Sep 14, 2018, 10:19 AM PDT.
// + error message "last payment declined"
// after a while a new error message shows with a "fix" call to action: "Payment declined, to keep your subscription, update payment method."
// json response is still the same.
// after some while, subscription is removed automatically, and empty json response.
