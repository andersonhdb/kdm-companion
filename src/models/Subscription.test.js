import { getSnapshot } from 'mobx-state-tree'
import { Subscription } from './Subscription'

// afterEach(() => {
//   jest.restoreAllMocks()
// })

it('.shouldCheck with old date', () => {
  let sub = Subscription.create({
    id: 'test',
    purchasedAt: Date.now() - 1000 * 60 * 60 * 24 * 31,
    checkedAt: Date.now() - 1000 * 60 * 60 * 24 * 31,
  })

  expect(sub.shouldCheck).toBeTruthy()
})

it('.shouldCheck with recent date', () => {
  let sub = Subscription.create({
    id: 'test',
    purchasedAt: Date.now() - 1000 * 60 * 60 * 24 * 29,
    checkedAt: Date.now() - 1000 * 60 * 60 * 24 * 29,
  })

  expect(sub.shouldCheck).toBeFalsy()
})
