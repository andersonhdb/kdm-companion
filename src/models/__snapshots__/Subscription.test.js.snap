// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.purchased handles the IAP response 1`] = `
Object {
  "appLastActiveAt": 0,
  "autoRenewing": true,
  "checkedAt": 1536939811199,
  "productId": "premium_subscription_1",
  "purchasedAt": 1536939811199,
  "response": "{\\"autoRenewingAndroid\\":true,\\"dataAndroid\\":\\"{orderId: GPA.3327-6252-8328-19548, packageName: 'com.github.ingoclaro.kdmcompanion', productId: 'premium_subscription_1', purchaseTime: 1536906424588, purchaseState: 0, purchaseToken: '[redacted]', autoRenewing: true}\\",\\"productId\\":\\"premium_subscription_1\\",\\"purchaseToken\\":\\"[redacted]\\",\\"signatureAndroid\\":\\"[redacted]\\",\\"transactionDate\\":\\"1536939811199\\",\\"transactionId\\":\\"GPA.3327-6252-8328-19548\\",\\"transactionReceipt\\":\\"[redacted]\\"}",
  "transactionId": "GPA.3327-6252-8328-19548",
}
`;
